File: solver/src/game/card.py
"""
Card representation for poker.
"""

from dataclasses import dataclass

from ..enums import Rank, Suit


@dataclass(frozen=True)
class Card:
    """
    Represents a playing card with type-safe rank and suit.
    Immutable to allow use in sets and as dict keys.
    """
    rank: Rank
    suit: Suit

    @property
    def notation(self) -> str:
        """Short notation for the card (e.g., 'As' for Ace of Spades)."""
        return f'{self.rank.symbol}{self.suit.symbol}'

    def __str__(self) -> str:
        return self.notation

    def __repr__(self) -> str:
        return f"Card('{self.notation}')"

    @classmethod
    def from_notation(cls, notation: str) -> 'Card':
        """
        Creates card from notation string (e.g., 'As', 'Th').
        Used internally for convenience, not exposed in API.
        """
        if len(notation) != 2:
            raise ValueError(f'Card notation must be 2 characters: {notation}')
        rank = Rank.from_symbol(notation[0])
        suit = Suit.from_symbol(notation[1])
        return cls(rank=rank, suit=suit)


def cards_from_string(s: str) -> list[Card]:
    """
    Parse multiple cards from a space-separated string.

    Args:
        s: Space-separated card notations (e.g., 'As Kh Qd')

    Returns:
        List of Card objects
    """
    return [Card.from_notation(c) for c in s.strip().split()]

File: backend/src/main/kotlin/com/mako/model/Card.kt
package com.mako.model

import com.mako.dto.CardDto
import com.mako.enums.Rank
import com.mako.enums.Suit

/**
 * Represents a playing card with type-safe rank and suit.
 *
 * @property rank Card rank (TWO through ACE)
 * @property suit Card suit (SPADES, HEARTS, DIAMONDS, CLUBS)
 */
data class Card(
    val rank: Rank,
    val suit: Suit
) {
    /**
     * Short notation for the card (e.g., "As" for Ace of Spades).
     */
    val notation: String get() = "${rank.symbol}${suit.symbol}"

    /**
     * Converts card to DTO for API response.
     * Uses display name for suit to match frontend expectations.
     */
    fun toDto(): CardDto = CardDto(
        rank = rank.symbol,
        suit = suit.displayName,
        display = notation
    )

    override fun toString(): String = notation

    companion object {
        /**
         * Creates card from notation string (e.g., "As", "Th").
         *
         * @param notation Two-character string: rank + suit
         * @return Card instance
         * @throws IllegalArgumentException if notation is invalid
         */
        fun fromNotation(notation: String): Card {
            require(notation.length == 2) {
                "Card notation must be 2 characters: $notation"
            }
            return Card(
                rank = Rank.fromSymbol(notation[0].toString()),
                suit = Suit.fromSymbol(notation[1].toString())
            )
        }
    }
}
